x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java

```diff
@@ -156,6 +151,153 @@ public void testIsServiceAccount() {
+        RealmRef authnRealmRef = randomRealmRef(true);
+        RealmRef lookupRealmRef = randomRealmRef(true);
+        // realm/token run-as
+        Authentication test = Authentication.newRealmAuthentication(randomUser(), authnRealmRef);
+        if (randomBoolean()) {
+            test = test.token();
+        }
+        test = test.runAs(randomUser(), lookupRealmRef);
+        if (randomBoolean()) {
+            test = test.token();
+        }
+        assertThat(test.isAssignedToDomain(), is(true));
+        assertThat(test.getDomain(), is(lookupRealmRef.getDomain()));
+        test = Authentication.newRealmAuthentication(randomUser(), randomRealmRef(false));
+        if (randomBoolean()) {
+            test = test.token();
+        }
+        test = test.runAs(randomUser(), lookupRealmRef);
+        if (randomBoolean()) {
+            test = test.token();
+        }
+        assertThat(test.isAssignedToDomain(), is(true));
+        assertThat(test.getDomain(), is(lookupRealmRef.getDomain()));
+        test = Authentication.newRealmAuthentication(randomUser(), authnRealmRef);
+        if (randomBoolean()) {
+            test = test.token();
+        }
+        test = test.runAs(randomUser(), randomRealmRef(false));
+        if (randomBoolean()) {
+            test = test.token();
+        }
+        assertThat(test.isAssignedToDomain(), is(false));
+        assertThat(test.getDomain(), nullValue());
+        test = Authentication.newRealmAuthentication(randomUser(), randomRealmRef(false));
+        if (randomBoolean()) {
+            test = test.token();
+        }
+        test = test.runAs(randomUser(), lookupRealmRef);
+        if (randomBoolean()) {
+            test = test.token();
+        }
+        assertThat(test.isAssignedToDomain(), is(true));
+        assertThat(test.getDomain(), is(lookupRealmRef.getDomain()));
+        // api key run-as
+        test = randomApiKeyAuthentication(randomUser(), randomAlphaOfLengthBetween(10, 20), Version.CURRENT);
+        if (randomBoolean()) {
+            test = test.token();
+        }
+        assertThat(test.isAssignedToDomain(), is(false));
+        assertThat(test.getDomain(), nullValue());
+        if (randomBoolean()) {
+            test = test.runAs(randomUser(), lookupRealmRef);
+            if (randomBoolean()) {
+                test = test.token();
+            }
+            assertThat(test.isAssignedToDomain(), is(true));
+            assertThat(test.getDomain(), is(lookupRealmRef.getDomain()));
+        } else {
+            test = test.runAs(randomUser(), randomRealmRef(false));
+            if (randomBoolean()) {
+                test = test.token();
+            }
+            assertThat(test.isAssignedToDomain(), is(false));
+            assertThat(test.getDomain(), nullValue());
+        }
+        // service account run-as
+        test = randomServiceAccountAuthentication();
+        if (randomBoolean()) {
+            test = test.token();
+        }
```

Comment:
Ha, yes, I hadn't notice it, thanks for catching; the condition to prevent creating a token by a service account is inside the transport action. My hunch was to look for it in the "create token" method.
I also tried it, and I see the failure is a `500` error.
I will raise a small PR for it.